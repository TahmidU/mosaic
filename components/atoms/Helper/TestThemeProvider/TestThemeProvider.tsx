import { ReactElement } from "react";
import { Theme } from "types/context";
import { ThemeProvider } from "styled-components";
import getTheme from "resources/themes";

interface TestThemeProviderProps {
  children: any;
  theme?: Theme;
}

export default function TestThemeProvider({
  children,
  theme = "light",
}: TestThemeProviderProps): ReactElement {
  return <ThemeProvider theme={getTheme(theme)}>{children}</ThemeProvider>;
}
