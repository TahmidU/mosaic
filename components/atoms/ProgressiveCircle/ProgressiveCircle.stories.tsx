import ProgressiveCircle from ".";
import { StoryUtils } from "resources/utils";

const payload = {
  title: StoryUtils.storyTitle("ATOMS", "ProgressiveCircle"),
  component: ProgressiveCircle,
};
export default payload;

export const Basic = (args: any) => (
  <ProgressiveCircle {...args} style={{ top: "100px", right: "50%" }} />
);
Basic.args = {
  width: 30,
  height: 30,
  radius: 30,
  strokeWidth: 4,
  duration: 1.75,
};
