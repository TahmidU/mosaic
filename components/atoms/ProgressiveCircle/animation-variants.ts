export function produceProgressiveCircleAnimVariant(
  radius: number,
  fillColourFrom: string,
  fillColourTo: string
) {
  return {
    show: {
      strokeDashoffset: 0,
      fill: fillColourTo,
    },
    hide: {
      strokeDashoffset: Math.ceil(2 * Math.PI * radius),
      fill: fillColourFrom,
    },
  };
}
