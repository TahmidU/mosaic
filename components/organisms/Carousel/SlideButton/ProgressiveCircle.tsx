import { AnimatedCircle } from "./styles";
import { ReactElement } from "react";

export default function ProgressiveCircle(): ReactElement {
  return (
    <svg
      viewBox="0 0 100 100"
      width={100}
      height={100}
      style={{
        position: "absolute",
        transform: "rotate(-90deg)",
        overflow: "visible",
        marginLeft: -100,
        top: "100px",
        right: "50%",
      }}
    >
      <AnimatedCircle
        cx="50"
        cy="50"
        r={100}
        strokeWidth={1}
        stroke="black"
        fill="transparent"
        strokeDashoffset={Math.ceil(2 * Math.PI * 100)}
        strokeDasharray={Math.abs(
          (Math.ceil(2 * Math.PI * 100) / 100) * (80 - 100)
        )}
      />
    </svg>
  );
}
