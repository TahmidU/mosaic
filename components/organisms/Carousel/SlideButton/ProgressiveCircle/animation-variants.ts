export function produceProgressiveCircleAnimVariant(radius: number) {
  return {
    show: {
      opacity: 1,
      strokeDashoffset: 0,
    },
    hide: {
      opacity: 0,
      strokeDashoffset: Math.ceil(2 * Math.PI * radius),
    },
  };
}
