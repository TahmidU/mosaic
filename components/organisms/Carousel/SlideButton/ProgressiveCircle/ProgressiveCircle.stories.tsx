import ProgressiveCircle from ".";
import { storyTitle } from "../../../../../resources/utils/story";

const payload = {
  title: storyTitle("ATOMS", "ProgressiveCircle"),
  component: ProgressiveCircle,
};
export default payload;

export const Basic = (args: any) => <ProgressiveCircle {...args} />;
Basic.args = {};
