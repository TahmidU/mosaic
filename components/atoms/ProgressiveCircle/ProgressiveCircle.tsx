import { AnimatedCircle, Container } from "./styles";

import { ReactElement } from "react";
import { produceProgressiveCircleAnimVariant } from "./animation-variants";

interface ProgressiveCircleProps {
  width: number;
  height: number;
  radius: number;
  strokeWidth: number;
  duration?: number;
  show?: boolean;
  fillColourFrom?: string;
  fillColourTo?: string;
  className?: string;
}

export default function ProgressiveCircle({
  width,
  height,
  radius,
  strokeWidth,
  duration = 1.75,
  show = true,
  fillColourFrom = "transparent",
  fillColourTo = "transparent",
  className,
}: ProgressiveCircleProps): ReactElement {
  return (
    <Container
      width={width}
      height={height}
      className={className}
      data-testid="ProgressiveCircleContainer"
    >
      <AnimatedCircle
        data-testid="ProgressiveCircleAnimatedCircle"
        cx={`${width / 2}`}
        cy={`${height / 2}`}
        r={radius}
        strokeWidth={strokeWidth}
        initial="hide"
        variants={produceProgressiveCircleAnimVariant(
          radius,
          fillColourFrom,
          fillColourTo
        )}
        animate={show ? "show" : "hide"}
        transition={{ duration: duration }}
      />
    </Container>
  );
}
