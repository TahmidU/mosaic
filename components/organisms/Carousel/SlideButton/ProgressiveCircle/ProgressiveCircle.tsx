import { AnimatedCircle, Container } from "./styles";

import { ReactElement } from "react";
import { produceProgressiveCircleAnimVariant } from "./animation-variants";

interface ProgressiveCircleProps {
  width: number;
  height: number;
  radius: number;
  strokeWidth: number;
  duration?: number;
  show?: boolean;
  className?: string;
}

export default function ProgressiveCircle({
  width,
  height,
  radius,
  strokeWidth,
  duration = 1.75,
  show = true,
  className,
}: ProgressiveCircleProps): ReactElement {
  return (
    <Container width={width} height={height} className={className}>
      <AnimatedCircle
        cx={`${width / 2}`}
        cy={`${height / 2}`}
        r={radius}
        strokeWidth={strokeWidth}
        initial="hide"
        variants={produceProgressiveCircleAnimVariant(radius)}
        animate={show ? "show" : "hide"}
        transition={{ duration: duration }}
      />
    </Container>
  );
}
