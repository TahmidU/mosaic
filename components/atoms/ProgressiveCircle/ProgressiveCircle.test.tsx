import "jest-styled-components";

import ProgressiveCircle from "./ProgressiveCircle";
import TestThemeProvider from "components/atoms/Helper/TestThemeProvider";
import getTheme from "resources/themes";
import { render } from "utils/test-config";
import renderer from "react-test-renderer";

describe("ProgressiveCircle", () => {
  test("Show state light theme styling", async () => {
    // Given
    const width = 30,
      height = 30,
      radius = 30,
      strokeWidth = 4,
      circumference = Math.ceil(2 * Math.PI * radius),
      lightTheme = getTheme("light");

    // When
    const { getByTestId } = render(
      <ProgressiveCircle
        show
        width={width}
        height={height}
        radius={radius}
        strokeWidth={strokeWidth}
      />
    );

    // Then
    await new Promise((r) => setTimeout(r, 1800));

    expect(getByTestId("ProgressiveCircleContainer")).toHaveStyleRule(
      "position",
      "absolute"
    );
    expect(getByTestId("ProgressiveCircleContainer")).toHaveStyleRule(
      "width",
      `${width}px`
    );
    expect(getByTestId("ProgressiveCircleContainer")).toHaveStyleRule(
      "height",
      `${height}px`
    );
    expect(getByTestId("ProgressiveCircleContainer")).toHaveStyleRule(
      "transform",
      "rotate(-90deg)"
    );
    expect(getByTestId("ProgressiveCircleContainer")).toHaveStyleRule(
      "overflow",
      "visible"
    );

    expect(getByTestId("ProgressiveCircleAnimatedCircle")).toHaveStyleRule(
      "position",
      "absolute"
    );
    expect(getByTestId("ProgressiveCircleAnimatedCircle")).toHaveStyleRule(
      "stroke",
      lightTheme.almostBlack
    );
    expect(getByTestId("ProgressiveCircleAnimatedCircle")).toHaveStyleRule(
      "stroke-width",
      `${strokeWidth}`
    );
    expect(getByTestId("ProgressiveCircleAnimatedCircle")).toHaveStyleRule(
      "stroke-dasharray",
      `${circumference}`
    );
    expect(getByTestId("ProgressiveCircleAnimatedCircle")).toHaveStyleRule(
      "stroke-linecap",
      "round"
    );
  });

  test("Hide state light theme styling", async () => {
    // Given
    const width = 30,
      height = 30,
      radius = 30,
      strokeWidth = 4,
      circumference = Math.ceil(2 * Math.PI * radius),
      strokeDashoffset = circumference,
      lightTheme = getTheme("light");

    // When
    const { getByTestId } = render(
      <ProgressiveCircle
        show
        width={width}
        height={height}
        radius={radius}
        strokeWidth={strokeWidth}
      />
    );

    const rendered = renderer.create(
      <TestThemeProvider>
        <ProgressiveCircle
          show
          width={width}
          height={height}
          radius={radius}
          strokeWidth={strokeWidth}
        />
      </TestThemeProvider>
    );
    const tree = rendered.toJSON();

    // Then
    await new Promise((r) => setTimeout(r, 1800));

    expect(getByTestId("ProgressiveCircleContainer")).toHaveStyleRule(
      "position",
      "absolute"
    );
    expect(getByTestId("ProgressiveCircleContainer")).toHaveStyleRule(
      "width",
      `${width}px`
    );
    expect(getByTestId("ProgressiveCircleContainer")).toHaveStyleRule(
      "height",
      `${height}px`
    );
    expect(getByTestId("ProgressiveCircleContainer")).toHaveStyleRule(
      "transform",
      "rotate(-90deg)"
    );
    expect(getByTestId("ProgressiveCircleContainer")).toHaveStyleRule(
      "overflow",
      "visible"
    );

    expect(getByTestId("ProgressiveCircleAnimatedCircle")).toHaveStyleRule(
      "position",
      "absolute"
    );
    expect(getByTestId("ProgressiveCircleAnimatedCircle")).toHaveStyleRule(
      "stroke",
      lightTheme.almostBlack
    );
    expect(getByTestId("ProgressiveCircleAnimatedCircle")).toHaveStyleRule(
      "stroke-width",
      `${strokeWidth}`
    );
    expect(getByTestId("ProgressiveCircleAnimatedCircle")).toHaveStyleRule(
      "stroke-dasharray",
      `${circumference}`
    );
    expect(rendered.root.findByType("circle").props.strokeDashoffset).toBe(
      strokeDashoffset
    ); //toHaveStyleRule does not recognise stroke-dashoffset
    expect(getByTestId("ProgressiveCircleAnimatedCircle")).toHaveStyleRule(
      "stroke-linecap",
      "round"
    );
  });

  test("Snapshot wait", async () => {
    // Given
    const width = 30,
      height = 30,
      radius = 30,
      strokeWidth = 4;

    // When
    const rendered = renderer.create(
      <TestThemeProvider>
        <ProgressiveCircle
          width={width}
          height={height}
          radius={radius}
          strokeWidth={strokeWidth}
        />
      </TestThemeProvider>
    );
    const tree = rendered.toJSON();

    // Then
    await new Promise((r) => setTimeout(r, 1800));
    expect(tree).toMatchSnapshot();
  });

  test("Snapshot non-wait", () => {
    // Given
    const width = 30,
      height = 30,
      radius = 30,
      strokeWidth = 4;

    // When
    const rendered = renderer.create(
      <TestThemeProvider>
        <ProgressiveCircle
          width={width}
          height={height}
          radius={radius}
          strokeWidth={strokeWidth}
        />
      </TestThemeProvider>
    );
    const tree = rendered.toJSON();

    // Then
    expect(tree).toMatchSnapshot();
  });
});
