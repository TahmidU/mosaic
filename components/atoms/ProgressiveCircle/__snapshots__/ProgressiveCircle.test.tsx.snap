// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressiveCircle Snapshot non-wait 1`] = `
.c0 {
  position: absolute;
  width: 30px;
  height: 30px;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  overflow: visible;
}

.c1 {
  position: absolute;
  stroke: #211f1b;
  stroke-width: 4;
  stroke-dasharray: 189;
  stroke-linecap: round;
}

<svg
  className="c0"
  data-testid="ProgressiveCircleContainer"
  height={30}
  width={30}
>
  <circle
    className="c1"
    cx="15"
    cy="15"
    data-testid="ProgressiveCircleAnimatedCircle"
    fill="transparent"
    r={30}
    strokeDashoffset={189}
    strokeWidth={4}
    style={Object {}}
  />
</svg>
`;

exports[`ProgressiveCircle Snapshot wait 1`] = `
.c0 {
  position: absolute;
  width: 30px;
  height: 30px;
  -webkit-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
  overflow: visible;
}

.c1 {
  position: absolute;
  stroke: #211f1b;
  stroke-width: 4;
  stroke-dasharray: 189;
  stroke-linecap: round;
}

<svg
  className="c0"
  data-testid="ProgressiveCircleContainer"
  height={30}
  width={30}
>
  <circle
    className="c1"
    cx="15"
    cy="15"
    data-testid="ProgressiveCircleAnimatedCircle"
    fill="transparent"
    r={30}
    strokeDashoffset={189}
    strokeWidth={4}
    style={Object {}}
  />
</svg>
`;
